File: src/main/java/labrary/digitaldepartment/Entity/Document.java
package labrary.digitaldepartment.Entity;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties; // 🌟 Добавили импорт
import jakarta.persistence.*;
import labrary.digitaldepartment.Enums.SyllabusStatus;
import lombok.Getter;
import lombok.Setter;
import lombok.NoArgsConstructor;
import java.util.List;

@Entity
@Getter @Setter
@NoArgsConstructor
@Table(name = "documents", schema = "business")
public class Document {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(cascade = CascadeType.PERSIST)
    @JoinColumn(name = "discipline_id", nullable = false)
    // 🌟 Игнорируем обратные ссылки внутри Discipline, если они там есть
    @JsonIgnoreProperties({"documents", "handler", "hibernateLazyInitializer"})
    private Discipline discipline;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    // 🌟 Игнорируем списки документов автора и его пароль ради безопасности
    @JsonIgnoreProperties({"documents", "password", "handler", "hibernateLazyInitializer"})
    private User author;

    private String academicYear;
    private Integer semester;

    @Enumerated(EnumType.STRING)
    private SyllabusStatus status;

    @OneToMany(mappedBy = "document", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<AcademicLoad> academicLoads;

    @OneToMany(mappedBy = "document", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<PaymentDetail> paymentDetails;

    @OneToMany(mappedBy = "document", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<WeeklyTopic> weeklyTopics;

    @OneToOne(mappedBy = "document", cascade = CascadeType.ALL, orphanRemoval = true)
    @PrimaryKeyJoinColumn
    @JsonIgnoreProperties("document") // 🌟 Защита от зацикливания с Силлабусом
    private Syllabus syllabus;

    public void setSyllabus(Syllabus syllabus) {
        if (syllabus == null) {
            if (this.syllabus != null) {
                this.syllabus.setDocument(null);
            }
        } else {
            syllabus.setDocument(this);
        }
        this.syllabus = syllabus;
    }

    @OneToMany(mappedBy = "document", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<CourseVolume> courseVolumes;
}

File: src/Front/src/app/Services/Document/Document.ts
import { User } from '../User/User';
import { Syllabus } from '../Syllabus/syllabus';
import { Department } from '../Department/department';

export enum DocumentStatus {
  DRAFT = 'DRAFT',
  UNDER_REVIEW = 'UNDER_REVIEW',
  SIGNED = 'SIGNED'
}

export interface Discipline {
  id?: number;
  name: string;
  courseCode: string;
  creditsEcts: number;
  department?: Department;
}

export interface AcademicLoad {
  id?: number;
  discipline: string;
  studentGroup: string;
  totalStreams: string;
  lectureHours: number;
  practiceHours: number;
  labHours: number;
  totalHours: number;
}

export interface PaymentDetail { // Убрал 's', чтобы совпадало с Java Entity
  id?: number;
  staffLoad: number;
  hourlyLoad: number;
}

export interface WeeklyTopic {
  id?: number;
  topicName: string; // добавьте поля согласно вашей Java WeeklyTopic Entity
  hours?: number;
}

export interface SyllabusDocument {
  id?: number;
  discipline: Discipline; // Убрал опциональность, так как в БД nullable=false
  status: DocumentStatus;
  academicYear: string;
  semester: number;
  author?: User;

  // Дочерние объекты
  syllabus?: Syllabus;
  academicLoads?: AcademicLoad[];
  paymentDetails?: PaymentDetail[];
  weeklyTopics?: WeeklyTopic[];
}
